### PR TITLE
stat-util: generalize is_* and verify_* handling

### PR DESCRIPTION
**Pull request by YHNdnzj**

### CODE REVIEW

File: src/basic/stat-util.c [hunk 1]
```diff
@@ -25,43 +25,131 @@
...
+
+        if (fstatat(fd, strempty(path), &st,
+                    (isempty(path) ? AT_EMPTY_PATH : 0) | (follow ? 0 : AT_SYMLINK_NOFOLLOW)) < 0)
                 return -errno;
 
-        return !!S_ISLNK(info.st_mode);
+        if (!verify)
+                return verify_func(&st) >= 0;
+
+        return verify_func(&st);
```
Comment (poettering) [inline src/basic/stat-util.c]: i'd write this as:

```c
r = verify_func(&st);

return verify ? r : r >= 0;
```

seems easier to read to me, and emphasizes that the difference really is just in the interpretation of the return value

File: src/basic/stat-util.c [hunk 2]
```diff
@@ -25,43 +25,131 @@
...
+}
+
+int is_dir_at(int fd, const char *path, bool follow) {
+        return verify_stat_at(fd, path, follow, stat_verify_directory, false);
+}
+
+int is_dir(const char *path, bool follow) {
+        assert(!isempty(path));
+        return is_dir_at(AT_FDCWD, path, follow);
+}
```
Comment (poettering) [inline src/basic/stat-util.c:101]: make this an inline func? but doesn't matter

File: src/basic/stat-util.c [hunk 3]
```diff
@@ -25,43 +25,131 @@
...
+        if (st->st_nlink <= 0)
+                return -EIDRM; /* recognizable error. */
+
+        return 0;
+}
+
+int fd_verify_linked(int fd) {
+        assert(fd >= 0);
+        return verify_stat_at(fd, NULL, false, stat_verify_linked, true);
+}
```
Comment (poettering) [inline src/basic/stat-util.c:132]: ditto

File: src/basic/stat-util.c (hunk 2 above)
Comment (YHNdnzj) [reply src/basic/stat-util.c:101]: Hmm, so some of these `is_` and `fd_verify_` functions are wrappers around `_at` variants, others are not. I think it's easier to manage if we place everything closely.

**Review by bluca — APPROVED** [review]

**Pull request merged**